File: 74. Search a 2D Matrix/74. Search a 2D Matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  const m = matrix.length;
  const n = matrix[0].length;

  let low = 0;
  let high = m * n - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const midI = Math.floor(mid / n);
    const midJ = mid % n;
    const midMatrix = matrix[midI][midJ];

    if (midMatrix === target) {
      return true;
    } else if (midMatrix > target) {
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  return false;
};

File: 74. Search a 2D Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        left = 0
        right = m * n - 1

        while left <= right:
            mid = (left + right) // 2

            i, j = mid // n, mid % n

            if matrix[i][j] == target:
                return True
            elif matrix[i][j] > target:
                right = mid - 1
            else:
                left = mid + 1

        return False
